CORDA-3291 `isKilled` flag and session errors for killed flows

lankydan opened the pull request:
## Summary

Two major improvements have been worked on:

- A new flag named `isKilled` has been added to `FlowLogic` to allow
developers to break out of loops without suspension points.
- Killed flows now send session errors to their counter parties allowing
their flows to also terminate without further coordination.

Achieving these changes required a __fundamental__ change to how flows are
killed as well as how they sleep.

## `isKilled` flag

The addition of `FlowLogic.isKilled` allows flows to check if the
current flow has been killed. They can then throw an exception to lead
to the flow's termination (following the standard error pathway). They
can also perform some extra logic or not throw an exception if they
really wanted to.

No matter what, once the flag is set, the flow will terminate. Due to
timing, a killed flow might successfully process its next suspension
event, but it will then process a killed transition and terminate.

## Send session errors when killing a flow

A flow will now send session errors to all of its counter parties. They
are transferred as `UnexpectedFlowEndException`s. This allows initiated
flows to handle these errors as they see fit, although they should
probably just terminate.

## How flows are now killed

### Before

Originally we were relying on Quasar to interrupt a flow's fiber, we
could then handle the resulting `InterruptedException`. The problem with
this solution is that it only worked when a flow was already suspended
or when a flow moved into suspension. Flows stuck in loops did not work.

### After

We now *do not* use Quasar to interrupt a flow's fiber. Instead, we
switch `FlowStateMachine.isKilled` to true and schedule a new event.
Any event that is processed after switching this flag will now cause a
`KilledFlowTransition`. This transition follows similar logic to how
error propagation works. Note, the extra event allows a suspended flow
to be killed without waiting for the event that it was _really_ waiting
for.

This allows a lot of the tidy up code in `StateMachineManager.killFlow`
to be removed as tidy up is executed as part of removing a flow.
Deleting a flow's checkpoint and releasing related soft locks is still
handled manually in case of infinite loops but also triggered as part
of the actions executed in a transition.

This required flow sleeping to be changed as we no longer rely on
quasar.

## How flows now sleep

The reliance on Quasar to make a flow sleep has been removed.

Instead, when a flow sleeps we create a `ScheduledFuture` that is
delayed for the requested sleep duration. When the future executes it
schedules a `WakeUpFromSleep` event that wakes up the flow... Duh.

`FlowSleepScheduler` handles the future logic. It also uses the same
scheduled thread pool that timed flows uses.

A future field was added to `StateMachineState`. This removes the 
need for concurrency control around flow sleeps as the code path does
not need to touch any concurrent data structures.

To achieve this:

- `StateMachineState.future` added as a `var`
- When the `ScheduledFuture` is created to wake up the flow the passed
in `StateMachineState` has its `future` value changed
- When resumed `future` and `isWaitingForFuture` are set to `null` and
`false` respectively
- When cancelling a sleeping flow, the `future` is cancelled and nulled
out. `isWaitingForFuture` is not changed since the flow is ending anyway
so really the value of the field is not important.

lankydan committed:
CORDA-3291 `isKilled` flag and session errors for killed flows

## Summary

Two major improvements have been worked on:

- A new flag named `isKilled` has been added to `FlowLogic` to allow
developers to break out of loops without suspension points.
- Killed flows now send session errors to their counter parties allowing
their flows to also terminate without further coordination.

Achieving these changes required a *fundamental* change to how flows are
killed as well as how they sleep.

## `isKilled` flag

The addition of `FlowLogic.isKilled` allows flows to check if the
current flow has been killed. They can then throw an exception to lead
to the flow's termination (following the standard error pathway). They
can also perform some extra logic or not throw an exception if they
really wanted to.

No matter what, once the flag is set, the flow will terminate. Due to
timing, a killed flow might successfully process its next suspension
event, but it will then process a killed transition and terminate.

## Send session errors when killing a flow

A flow will now send session errors to all of its counter parties. They
are transferred as `UnexpectedFlowEndException`s. This allows initiated
flows to handle these errors as they see fit, although they should
probably just terminate.

## How flows are now killed

### Before

Originally we were relying on Quasar to interrupt a flow's fiber, we
could then handle the resulting `InterruptedException`. The problem with
this solution is that it only worked when a flow was already suspended
or when a flow moved into suspension. Flows stuck in loops did not work.

### After

We now *do not* use Quasar to interrupt a flow's fiber. Instead, we
switch `FlowStateMachine.isKilled` to true and schedule a new event.
Any event that is processed after switching this flag will now cause a
`KilledFlowTransition`. This transition follows similar logic to how
error propagation works. Note, the extra event allows a suspended flow
to be killed without waiting for the event that it was _really_ waiting
for.

This allows a lot of the tidy up code in `StateMachineManager.killFlow`
to be removed as tidy up is executed as part of removing a flow.
Deleting a flow's checkpoint and releasing related soft locks is still
handled manually in case of infinite loops.

This required flow sleeping to be changed as we not longer rely on
quasar.

## How flows now sleep

The reliance on Quasar to make a flow sleep has been removed.

Instead, when a flow sleeps we create a `ScheduledFuture` that is
delayed for the requested sleep duration. When the future executes it
schedules a `WakeUpFromSleep` event that wakes up the flow... Duh.

`FlowSleepScheduler` handles the future logic. It also uses the same
scheduled thread pool that timed flows uses.

lankydan committed:
CORDA-3291 Detekt fixes

mnesbit commented on core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt:
```diff
@@ -127,6 +127,31 @@ abstract class FlowLogic<out T> {
+     *
+     * Note, once the [isKilled] flag is set to `true` the flow will terminate once it reaches its next suspension point. Therefore, it is
+     * possible to write a flow that does not interact with the [isKilled] flag.
+     */
```
Really the statement should be clearer. The flow may terminate on the next call to any API method marked as @Suspendable. We don't really define when 'suspension points' happen.

Also can we confirm that flow.sleep will die immediately even if below the 'suspension' minimum duration?

lankydan replied:
Yeah that makes some sense.

I'll add an assertion to my tests.

mnesbit commented on core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt:
````diff
@@ -570,6 +595,46 @@ abstract class FlowLogic<out T> {
+     * ```
+     *
+     * See the [isKilled] property for more information.
+     */
````
Would it be better to implement a runInterruptably method that takes a loop lambda and possibly a terminaion message

lankydan replied:
This is something that I considered. 

Kotlin uses a `isActive` flag for their coroutines in the same way that we have used `isKilled`, so I don't think what we have done so far is different from other platforms.

The main reason I added these helpers is to allow developers to continue writing code similar to how they normally do. What I mean is, is that it might be easier to discover `isKilled` rather than a new way to do loops.

I am fine adding a function for it.  But there might be a wide surface area. `forEach`, `for` loops, `while` loops, `map` calls etc.

mnesbit reviewed (commented):
Seems a little late to add this API and whenever it goes in we must have a new platform version.

lankydan commented:
> Seems a little late to add this API and whenever it goes in we must have a new platform version.

Without this API we will never be able to break out of infinite/long running loops. Obviously it does require the developer to leverage it.

lankydan committed:
CORDA-3291 Improve `isKilled` KDoc

lankydan commented:
One thing I want to highlight is that we are communicating killed flows to counter party nodes via an `UnexpectedFlowEndException`. 

I discussed with @rick-r3 about adding to the message that is was due to a killed flow but from looking at the code, I don't think that is possible. `ErrorSessionMessage` only allows `FlowException`s to be attached and changing this would change the network version? 

The other option is to send a `FlowException` (`KilledFlowException`) to counter parties.

Basically, what it comes to, do we want to let counter parties know when a flow it is interacting with has been killed or should this just look like an internal error.

@mnesbit @rick-r3

lankydan committed:
CORDA-3291 Assert that a sleeping flow is killed quickly

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/transitions/KilledFlowTransition.kt:
```diff
@@ -0,0 +1,117 @@
+                actions.add(Action.CreateTransaction)
+            }
+
+            // The checkpoint and soft locks are removed directly in [StateMachineManager.killFlow]
```
I think we should also remove the checkpoint and release soft locks here.  We don't want the other attempt in `killFlow` to result in some weird edge case where we have persisted e.g. de-duplication facts and other database changes and committed because flow exited through here but `killFlow` didn't complete and not deleted the checkpoint.  The node is then restarted and the flow retries from last checkpoint but we committed the database changes (except removal of checkpoint).

rick-r3 reviewed (commented):
Looks good. Have one request re: the kill flow transition.  Hopefully it doesn't cause to much change to any tests.

rick-r3 commented:
BTW, we did increment the platform version.  Now at 7. (4.4 was 6)

lankydan committed:
CORDA-3291 Fix hanging test

kyriathar commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowSleepScheduler.kt:
```diff
@@ -0,0 +1,68 @@
+        return scheduledExecutor.schedule<Unit>(
+            {
+                log.debug { "Scheduling flow wake up event for flow ${flow.fiber.id}" }
+                innerState.sleepingFlows.remove(flow.fiber.id)
```
The code on this lambda is going to be executed by another thread therefore , maybe, the `innerState.sleepingFlows` should be called within a ` mutex.locked { }` block. @rick-r3 's input would be needed here also i think.

As discussed offline with @lankydan , some possible solutions would be:
1. Calling this with an action upon having a  `WakeUpFromSleep` event.
2. Passing in to `FlowSleepScheduler`  a reference to the `StateMachineManager` and add a generic lock-grabbing method in there to execute passed-in functions with the mutex acquired. Or add a remove sleeping flow method to the `StateMachineManager`.

lankydan committed:
CORDA-3291 Do not `unpark` fiber if flow is aborted

lankydan committed:
CORDA-3291 Keep track of future in `StateMachineState`

Instead of relying on a map containing all the sleeping flows and
their futures, keep the future on the `StateMachineState` and manage it
directly there.

This removes the need for concurrency control around flow sleeps as the
code path does not need to touch any concurrent data structures.

To achieve this:

- `StateMachineState.future` added as a `var`
- `StateMachineState.isWaitingForFuture` __changed__ to a `var`
- When the `ScheduledFuture` is created to wake up the flow the passed
in `StateMachineState` has its `future` value changed
- When resumed `future` and `isWaitingForFuture` are set to `null` and
`false` respectively
- When cancelling a sleeping flow, the `future` is cancelled and nulled
out and `isWaitingForFuture` is set to false (this only happens if the
flow was actually sleeping)

lankydan committed:
CORDA-3291 Remove warning

lankydan committed:
CORDA-3291 Change `isWaitingForFuture` back to a `val`

lankydan committed:
CORDA-3291 Fix after rebase

lankydan commented:
Let me merge whenever this is ready 😄

lankydan committed:
CORDA-3291 Remove checkpoint and soft locks in killed transition

lankydan committed:
CORDA-3291 Remove warning

lankydan committed:
CORDA-3291 Cancel sleep future when retrying a flow

lankydan committed:
CORDA-3291 Improve `KillFlowTest`

lankydan committed:
CORDA-3291 Remove hanging test and move tests into `KillFlowTest`
